### PR TITLE
Refactors XDP to Separate Interface Queues and Rules

### PR DESCRIPTION
**Pull request by nibanks**
The first step in supporting per-flow rules with XDP/DPDK layers; which will happen dynamically as a result of new sockets getting created by the QUIC layer.

### CODE REVIEW

File: src/platform/datapath_raw_xdp.c [hunk 1]
```diff
@@ -56,10 +60,17 @@ typedef struct _XDP_QUEUE {
     CXPLAT_LIST_ENTRY TxQueue;
 } XDP_QUEUE;
 
+typedef struct XDP_FLOW {
```
Comment (csujedihy) [inline src/platform/datapath_raw_xdp.c:63]: I am little confused. What does a flow mean? Is it related to any design decisions we chose?

Comment (csujedihy) [reply src/platform/datapath_raw_xdp.c:63]: So, now a datapath can have multiple interfaces and each interface can have multiple flows and each flow can have multiple queues?

File: src/platform/datapath_raw_xdp.c [hunk 2]
```diff
@@ -837,22 +846,32 @@ CxPlatDpRawInitialize(
...
-                        LibraryErrorStatus,
-                        "[ lib] ERROR, %u, %s.",
-                        Status,
-                        "CxPlatDpRawInterfaceInitialize");
                     CxPlatFree(Interface, IF_TAG);
                     continue;
                 }
+
+                // TODO - Don't hard code flows. Dynamically create them in response
+                // to sockets getting created. Will be done in follow up PR.
```
Comment (csujedihy) [inline src/platform/datapath_raw_xdp.c:867]: what type of sockets are you referring to? XSKs or CXPLAT_SOCKETs?

File: src/platform/datapath_raw_xdp.c (hunk 1 above)
Comment (csujedihy) [reply src/platform/datapath_raw_xdp.c:63]: What is the purpose of this extra layer of abstration?

File: src/platform/datapath_raw_xdp.c [hunk 3]
```diff
@@ -56,10 +60,17 @@ typedef struct _XDP_QUEUE {
     CXPLAT_LIST_ENTRY TxQueue;
 } XDP_QUEUE;
 
+typedef struct XDP_FLOW {
+    const XDP_INTERFACE* Interface;
+    CXPLAT_LIST_ENTRY Link;
+    XDP_QUIC_FLOW Descriptor;
+    XDP_QUEUE Queues[0];
```
Comment (csujedihy) [inline src/platform/datapath_raw_xdp.c:67]: what it this? It's not defined anywhere.

Comment (csujedihy) [reply src/platform/datapath_raw_xdp.c:67]: Btw, it doesn't seem to compile.

File: src/platform/datapath_raw_xdp.c (hunk 2 above)
Comment (nibanks) [reply src/platform/datapath_raw_xdp.c:867]: Eventually, they will be equivalent, but right now, XSKs.

File: src/platform/datapath_raw_xdp.c (hunk 3 above)
Comment (nibanks) [reply src/platform/datapath_raw_xdp.c:67]: Yes it does: https://dev.azure.com/ms/msquic/_build/results?buildId=249521&view=results

File: src/platform/datapath_raw_xdp.c (hunk 1 above)
Comment (nibanks) [reply src/platform/datapath_raw_xdp.c:63]: Every new socket that creates will need to create a new queue + rule to receive only that traffic. See `XDP_QUIC_FLOW`.

Comment (csujedihy) [reply src/platform/datapath_raw_xdp.c:63]: There's only one hit for XDP_QUIC_FLOW, which is in the struct definition of XDP_FLOW.

Every socket means literally every socket? Like for both client and listener? CxPlatXdpWorkerThread loops over each flow, which is going to be a problem when the number of flows is huge.

File: src/platform/datapath_raw_xdp.c (hunk 3 above)
Comment (csujedihy) [reply src/platform/datapath_raw_xdp.c:67]: I can't find the definition of XDP_QUIC_FLOW. Where is it?

Comment (nibanks) [reply src/platform/datapath_raw_xdp.c:67]: It's in the latest XDP header (program.h I beleive). You need to update your XDP to have it.

File: src/platform/datapath_raw_xdp.c (hunk 1 above)
Comment (nibanks) [reply src/platform/datapath_raw_xdp.c:63]: For listeners, there's one socket per local UDP port used. For clients, yes, there is one per client connection.

File: src/platform/datapath_raw_xdp.c (hunk 3 above)
Comment (csujedihy) [reply src/platform/datapath_raw_xdp.c:67]: ah I see.

File: src/platform/datapath_raw_xdp.c (hunk 1 above)
Comment (nibanks) [reply src/platform/datapath_raw_xdp.c:63]: Yi and I discussed this a bit and decided on a slight change of direction. Please wait for the next commit.

**Review by csujedihy — APPROVED** [review]

**Pull request merged**